File: librefit-service/src/main/kotlin/io/tohuwabohu/calc/Wizard.kt
package io.tohuwabohu.calc

import io.smallrye.mutiny.Uni
import io.tohuwabohu.crud.error.ErrorDescription
import io.tohuwabohu.crud.error.ValidationError
import io.tohuwabohu.crud.util.OneOfFloat
import jakarta.enterprise.context.ApplicationScoped
import jakarta.inject.Inject
import jakarta.validation.Validator
import jakarta.validation.constraints.Max
import jakarta.validation.constraints.Min
import jakarta.validation.constraints.NotNull
import java.time.LocalDate
import java.time.temporal.ChronoUnit
import kotlin.math.pow
import kotlin.math.round

@ApplicationScoped
class Wizard {
    @Inject
    private lateinit var validator: Validator

    fun validate(wizardInput: WizardInput) {
        val violations = validator.validate(wizardInput)

        if (violations.isNotEmpty()) {
            val errors = violations.map { violation ->
                ErrorDescription(violation.propertyPath.filterNotNull()[0].name, violation.message)
            }

            throw ValidationError(errors)
        }
    }

    fun validate(wizardTargetWeightInput: WizardTargetWeightInput) {
        val violations = validator.validate(wizardTargetWeightInput)

        if (violations.isNotEmpty()) {
            val errors = violations.map { violation ->
                ErrorDescription(violation.propertyPath.filterNotNull()[0].name, violation.message)
            }

            throw ValidationError(errors)
        }
    }

    fun validate(wizardTargetDateInput: WizardTargetDateInput) {
        val violations = validator.validate(wizardTargetDateInput)

        if (violations.isNotEmpty()) {
            val errors = violations.map { violation ->
                ErrorDescription(violation.propertyPath.filterNotNull()[0].name, violation.message)
            }

            throw ValidationError(errors)
        }
    }

    /**
     * Calculation of basic metabolic rate (BMR) and total daily energy expenditure (TDEE) based on the Harris-Benedict
     * formula.
     */
    fun calculate(wizardInput: WizardInput): Uni<WizardResult> {
        validate(wizardInput)
        
        val wizardResult = WizardResult()

        val targetBmiRange = calculateTargetBmi(wizardInput.age)

        wizardResult.bmr = calculateBmr(wizardInput.sex, wizardInput.weight.toFloat(), wizardInput.height.toFloat(), wizardInput.age.toFloat())
        wizardResult.tdee = calculateTdee(wizardInput.activityLevel, wizardResult.bmr)
        wizardResult.deficit = calculateDeficit(wizardInput.weeklyDifference.toFloat())
        wizardResult.target = calculateTarget(wizardInput.calculationGoal, wizardResult.tdee, wizardResult.deficit)
        wizardResult.bmi = calculateBmi(wizardInput.weight, wizardInput.height)
        wizardResult.bmiCategory = calculateBmiCategory(wizardInput.sex, wizardResult.bmi)
        wizardResult.targetBmi = (targetBmiRange.first + targetBmiRange.last) / 2
        wizardResult.targetBmiLower = targetBmiRange.first
        wizardResult.targetBmiUpper = targetBmiRange.last
        wizardResult.targetWeight = calculateTargetWeight(targetBmiRange, wizardInput.height.toFloat())
        wizardResult.targetWeightLower = calculateTargetWeightLower(targetBmiRange, wizardInput.height.toFloat())
        wizardResult.targetWeightUpper = calculateTargetWeightUpper(targetBmiRange, wizardInput.height.toFloat())
        wizardResult.recommendation = calculateRecommendation(wizardResult.bmiCategory)

        calculateDurationDaysTotal(wizardInput, wizardResult)

        return Uni.createFrom().item(wizardResult)
    }

    internal fun calculateBmr(sex: CalculationSex, weight: Float, height: Float, age: Float): Float = when (sex) {
        CalculationSex.MALE -> round(66 + (13.7 * weight) + (5 * height) - (6.8 * age)).toFloat()
        CalculationSex.FEMALE -> round(655 + (9.6 * weight) + (1.8 * height) - (4.7 * age)).toFloat()
    }

    internal fun calculateTdee(activityLevel: Float, bmr: Float) = round(activityLevel * bmr)

    internal fun calculateDeficit(weeklyDifference: Float) = weeklyDifference / 10 * 7000 / 7

    internal fun calculateTarget(calculationGoal: CalculationGoal, tdee: Number, deficit: Float): Float = when (calculationGoal) {
        CalculationGoal.GAIN -> tdee.toFloat() + deficit
        CalculationGoal.LOSS -> tdee.toFloat() - deficit;
    }

    internal fun calculateBmi(weight: Number, height: Number) = round(weight.toFloat() / ((height.toFloat() / 100).pow(2)))

    internal fun calculateBmiCategory(sex: CalculationSex, bmi: Float): BmiCategory = when (sex) {
        CalculationSex.FEMALE -> {
            when (bmi) {
                in 0f..18f -> BmiCategory.UNDERWEIGHT
                in 19f..24f -> BmiCategory.STANDARD_WEIGHT
                in 25f..30f -> BmiCategory.OVERWEIGHT
                in 31f..40f -> BmiCategory.OBESE
                else -> BmiCategory.SEVERELY_OBESE
            }
        }

        CalculationSex.MALE -> {
            when (bmi) {
                in 0f..19f -> BmiCategory.UNDERWEIGHT
                in 20f..25f -> BmiCategory.STANDARD_WEIGHT
                in 26f..30f -> BmiCategory.OVERWEIGHT
                in 31f..40f -> BmiCategory.OBESE
                else -> BmiCategory.SEVERELY_OBESE
            }
        }
    }

    internal fun calculateTargetBmi(age: Number): IntRange = when (age) {
        in 19..24 -> 19..24
        in 25..34 -> 20..25
        in 35..44 -> 21..26
        in 45..54 -> 22..27
        in 55..64 -> 23..28
        else -> 24..29
    }

    internal fun calculateTargetWeight(targetBmi: IntRange, height: Float) =
        round(((targetBmi.first + targetBmi.last).toFloat() / 2) * (height / 100).pow(2))

    internal fun calculateTargetWeightLower(targetBmi: IntRange, height: Float) =
        round(((targetBmi.first).toFloat()) * (height / 100).pow(2))

    internal fun calculateTargetWeightUpper(targetBmi: IntRange, height: Float) =
        round(((targetBmi.last).toFloat()) * (height / 100).pow(2))

    internal fun calculateDurationDaysTotal(wizardInput: WizardInput, wizardResult: WizardResult) {
        when (wizardInput.calculationGoal) {
            CalculationGoal.GAIN -> {
                wizardResult.durationDays = calculateDurationDays(wizardResult.targetWeight, wizardInput.weight.toFloat(), wizardResult.deficit)
                wizardResult.durationDaysUpper = calculateDurationDays(wizardResult.targetWeightUpper, wizardInput.weight.toFloat(), wizardResult.deficit)
                wizardResult.durationDaysLower = calculateDurationDays(wizardResult.targetWeightLower, wizardInput.weight.toFloat(), wizardResult.deficit)
            }

            CalculationGoal.LOSS -> {
                wizardResult.durationDays = calculateDurationDays(wizardInput.weight.toFloat(), wizardResult.targetWeight, wizardResult.deficit)
                wizardResult.durationDaysUpper = calculateDurationDays(wizardInput.weight.toFloat(), wizardResult.targetWeightUpper, wizardResult.deficit)
                wizardResult.durationDaysLower = calculateDurationDays(wizardInput.weight.toFloat(), wizardResult.targetWeightLower, wizardResult.deficit)
            }
        }
    }

    internal fun calculateDurationDays(weight: Float, targetWeight: Float, deficit: Float) = ((weight - targetWeight) * 7000) / deficit

    internal fun calculateRecommendation(bmiCategory: BmiCategory): WizardRecommendation = when (bmiCategory) {
        BmiCategory.STANDARD_WEIGHT -> WizardRecommendation.HOLD
        BmiCategory.UNDERWEIGHT -> WizardRecommendation.GAIN
        BmiCategory.OVERWEIGHT -> WizardRecommendation.LOSE
        BmiCategory.OBESE -> WizardRecommendation.LOSE
        BmiCategory.SEVERELY_OBESE -> WizardRecommendation.LOSE
    }

    /**
     * Calculation of realistic weight loss/gain goals until a specific date occurs.
     */
    fun calculateForTargetDate(input: WizardTargetDateInput): Uni<WizardTargetDateResult> {
        validate(input)

        val today = LocalDate.now()

        val daysBetween = ChronoUnit.DAYS.between(today, input.targetDate)

        val currentBmi = calculateBmi(input.currentWeight, input.height)
        val currentClassification = calculateBmiCategory(input.sex, currentBmi)

        val rates: List<Int> = listOf(100, 200, 300, 400, 500, 600, 700)
        val multiplier = if (input.calculationGoal === CalculationGoal.LOSS) -1 else 1
        val obeseList = listOf(BmiCategory.OBESE, BmiCategory.SEVERELY_OBESE)

        var resultByRate: Map<Int, WizardResult> = rates.associateWith {
            val weight = (input.currentWeight.toFloat() + Math.round((multiplier * (daysBetween * it) / 7000).toDouble()))

            val result = WizardResult()
            result.bmi = calculateBmi(weight, input.height)
            result.bmiCategory = calculateBmiCategory(input.sex, result.bmi)
            result.targetWeight = weight

            result
        }.filter { (_, result) ->
            /**
             * filter non-recommendable results:
             * - weight loss should not lead to an underweight BMI
             * - weight gain should not lead to an obese or severally obese BMI
             * - no weight gain option current BMI is of obese or severally obese category
             */
            when (input.calculationGoal) {
                CalculationGoal.LOSS -> {
                    result.bmiCategory != BmiCategory.UNDERWEIGHT
                }
                else -> {
                    !obeseList.contains(result.bmiCategory) || obeseList.contains(currentClassification)
                }
            }
        }

        // calculate a custom rate to present at least one viable result
        if (resultByRate.isEmpty()) {
            val targetBmi = calculateTargetBmi(input.age)
            val targetWeight = calculateTargetWeight(targetBmi, input.height.toFloat())

            val difference = when (input.calculationGoal) {
                CalculationGoal.GAIN -> {
                    Math.round(targetWeight - input.currentWeight.toFloat())
                }
                CalculationGoal.LOSS -> {
                    Math.round(input.currentWeight.toFloat() - targetWeight)
                }
            }

            val rate = Math.round(difference.toFloat() * 7000 / daysBetween)

            val result = WizardResult()
            result.bmi = calculateBmi(targetWeight, input.height)
            result.bmiCategory = calculateBmiCategory(input.sex, result.bmi)
            result.targetWeight = targetWeight

            resultByRate = mapOf(rate to result)
        }

        return Uni.createFrom().item(WizardTargetDateResult(
            resultByRate
        ))
    }

    /**
     * Calculation of length until a specific weight can be obtained.
     */
    fun calculateForTargetWeight(input: WizardTargetWeightInput): Uni<WizardTargetWeightResult> {
        validate(input)

        val currentBmi = calculateBmi(input.currentWeight, input.height)
        val targetWeightBmi = calculateBmi(input.targetWeight, input.height)

        val currentClassification = calculateBmiCategory(input.sex, currentBmi);
        val targetClassification = calculateBmiCategory(input.sex, targetWeightBmi)

        val difference = (if (input.targetWeight > input.currentWeight) {
            input.targetWeight - input.currentWeight
        } else {
            input.currentWeight - input.targetWeight
        }).toDouble()

        val warning = targetClassification == BmiCategory.UNDERWEIGHT || targetClassification == BmiCategory.OBESE || targetClassification == BmiCategory.SEVERELY_OBESE
        var message = ""

        if (targetClassification == BmiCategory.UNDERWEIGHT) {
            if (input.targetWeight < input.currentWeight) {
                message = if (currentClassification == BmiCategory.UNDERWEIGHT) {
                    "Your target weight is even lower that your current weight. Considering that you are currently underweight, I cannot recommend you to proceed."
                } else {
                    "Your target weight will classify you as underweight. You should revisit your choice."
                }
            }
        } else if (targetClassification == BmiCategory.OBESE) {
            if (input.targetWeight > input.currentWeight) {
                message = if (currentClassification == BmiCategory.OBESE) {
                    "Your target weight is even higher that your current weight. Considering that you are currently obese, I cannot recommend you to proceed."
                } else {
                    "Your target weight will classify you as obese. You should revisit your choice."
                }
            }
        } else if (targetClassification == BmiCategory.SEVERELY_OBESE) {
            if (input.targetWeight > input.currentWeight) {
                message = if(currentClassification == BmiCategory.SEVERELY_OBESE) {
                    "Your target weight is even higher that your current weight. Considering that you are currently severely obese, I cannot recommend you to proceed."
                } else {
                    "Your target weight will classify you as severely obese. You should revisit your choice."
                }
            }
        }

        var datePerRate: Map<Int, LocalDate> = emptyMap()

        if ("" == message) {
            val rates: List<Int> = listOf(100, 200, 300, 400, 500, 600, 700)
            datePerRate = rates.associateWith { input.startDate.plusDays(Math.round(difference * 7000 / it)) }
        }

        return Uni.createFrom().item(WizardTargetWeightResult(datePerRate, targetClassification, warning, message))
    }
}

data class WizardInput (
    @field:Min(value = 18, message = "Your age must be between 18 and 99 years.")
    @field:Max(value = 99, message = "Your age must be between 18 and 99 years.")
    val age: Number,

    val sex: CalculationSex,

    @field:Min(value = 30, message = "Please provide a weight between 30kg and 300kg.")
    @field:Max(value = 300, message = "Please provide a weight between 30kg and 300kg.")
    val weight: Number,

    @field:Min(value = 30, message = "Please provide a height between 100cm and 220cm.")
    @field:Max(value = 300, message = "Please provide a height between 100cm and 220cm.")
    val height: Number,

    @OneOfFloat(value = [1f, 1.25f, 1.5f, 1.75f, 2f], message = "Please enter a valid activity level.")
    val activityLevel: Float,

    @field:Min(value = 0, message = "Your weekly difference must be between 0 and 0.7kg.")
    @field:Max(value = 7, message = "Your weekly difference must be between 0 and 0.7kg.")
    val weeklyDifference: Number,
    val calculationGoal: CalculationGoal,
)

data class WizardResult (
    var bmr: Float = 0f,
    var tdee: Number = 0,
    var deficit: Float = 0f,
    var target: Float = 0f,
    var bmi: Float = 0f,
    var bmiCategory: BmiCategory = BmiCategory.STANDARD_WEIGHT,
    var recommendation: WizardRecommendation = WizardRecommendation.HOLD,
    var targetBmi: Int = 23,
    var targetBmiUpper: Int = 24,
    var targetBmiLower: Int = 20,
    var targetWeight: Float = 0f,
    var targetWeightUpper: Float = 0f,
    var targetWeightLower: Float = 0f,
    var durationDays: Number = 0,
    var durationDaysUpper: Number = 0,
    var durationDaysLower: Number = 0
)

data class WizardTargetWeightInput (
    @field:Min(value = 18, message = "Your age must be between 18 and 99 years.")
    @field:Max(value = 99, message = "Your age must be between 18 and 99 years.")
    val age: Number,

    val sex: CalculationSex,

    @field:Min(value = 30, message = "Please provide a weight between 30kg and 300kg.")
    @field:Max(value = 300, message = "Please provide a weight between 30kg and 300kg.")
    val currentWeight: Float,

    @field:Min(value = 100, message = "Please provide a height between 100cm and 220cm.")
    @field:Max(value = 300, message = "Please provide a height between 100cm and 220cm.")
    val height: Number,

    @field:Min(value = 30, message = "Please provide a weight between 30kg and 300kg.")
    @field:Max(value = 300, message = "Please provide a weight between 30kg and 300kg.")
    val targetWeight: Float,

    val startDate: LocalDate
)

data class WizardTargetWeightResult (
    var datePerRate: Map<Int, LocalDate> = emptyMap(),
    var targetClassification: BmiCategory = BmiCategory.STANDARD_WEIGHT,
    val warning: Boolean,
    val message: String
)

data class WizardTargetDateInput (
    @field:Min(value = 18, message = "Your age must be between 18 and 99 years.")
    @field:Max(value = 99, message = "Your age must be between 18 and 99 years.")
    val age: Number,

    val sex: CalculationSex,

    @field:Min(value = 30, message = "Please provide a weight between 30kg and 300kg.")
    @field:Max(value = 300, message = "Please provide a weight between 30kg and 300kg.")
    val currentWeight: Number,

    @field:Min(value = 100, message = "Please provide a height between 100cm and 220cm.")
    @field:Max(value = 300, message = "Please provide a height between 100cm and 220cm.")
    val height: Number,

    val calculationGoal: CalculationGoal,

    @field:NotNull
    val targetDate: LocalDate
)


data class WizardTargetDateResult (
    var resultByRate: Map<Int, WizardResult> = emptyMap(),
)

enum class CalculationGoal {
    GAIN,
    LOSS
}

enum class CalculationSex {
    MALE,
    FEMALE
}

enum class BmiCategory {
    UNDERWEIGHT,
    STANDARD_WEIGHT,
    OVERWEIGHT,
    OBESE,
    SEVERELY_OBESE,
}

enum class WizardRecommendation {
    HOLD,
    LOSE,
    GAIN
}

File: src-tauri/src/calc/wizard.rs
use crate::crud::db::model::{CalorieTarget, NewCalorieTarget, NewWeightTarget, NewWeightTracker, WeightTarget, WeightTracker};
use chrono::{Duration, NaiveDate, Utc};
use serde::{Deserialize, Serialize};
use std::collections::HashMap;
use std::ops::RangeInclusive;
use validator::{Validate, ValidationError};

#[derive(Serialize, Deserialize, Debug)]
#[serde(rename_all = "camelCase")]
pub struct Wizard {
    pub calorie_target: NewCalorieTarget,
    pub weight_target: NewWeightTarget,
    pub weight_tracker: NewWeightTracker,
}

#[derive(Validate, Deserialize, Debug)]
#[serde(rename_all = "camelCase")]
pub struct WizardInput {
    #[validate(range(min = 18, max = 99))]
    age: i32,
    sex: CalculationSex,
    #[validate(range(min = 30f32, max = 300f32))]
    weight: f32,
    #[validate(range(min = 100f32, max = 300f32))]
    height: f32,
    #[validate(custom(function = "validate_activity_level"))]
    activity_level: f32,
    #[validate(range(min = 0, max = 7))]
    weekly_difference: i32,
    calculation_goal: CalculationGoal,
}

#[derive(Serialize, Debug)]
#[serde(rename_all = "camelCase")]
pub struct WizardResult {
    bmr: f32,
    tdee: f32,
    deficit: f32,
    target: f32,
    bmi: f32,
    bmi_category: BmiCategory,
    recommendation: WizardRecommendation,
    target_bmi: i32,
    target_bmi_upper: i32,
    target_bmi_lower: i32,
    target_weight: f32,
    target_weight_upper: f32,
    target_weight_lower: f32,
    duration_days: i32,
    duration_days_upper: i32,
    duration_days_lower: i32,
}

impl WizardResult {
    pub fn specific(
        bmi: &f32,
        bmi_category_clone: BmiCategory,
        target_weight: &f32,
    ) -> Self {
        WizardResult {
            bmr: 0.0,
            tdee: 0.0,
            deficit: 0.0,
            target: 0.0,
            bmi: *bmi,
            bmi_category: bmi_category_clone,
            target_weight: *target_weight,
            target_bmi: 0,
            target_bmi_upper: 0,
            target_bmi_lower: 0,
            target_weight_upper: 0.0,
            target_weight_lower: 0.0,
            duration_days: 0,
            duration_days_upper: 0,
            duration_days_lower: 0,
            recommendation: WizardRecommendation::HOLD,
        }
    }
}

#[derive(Deserialize, Debug)]
#[serde(rename_all = "camelCase")]
pub struct WizardTargetWeightInput {
    age: i32,
    sex: CalculationSex,
    current_weight: f32,
    height: f32,
    target_weight: f32,
    #[serde(deserialize_with = "crate::crud::serde::date::deserialize")]
    start_date: NaiveDate,
}

#[derive(Serialize, Debug)]
#[serde(rename_all = "camelCase")]
pub struct WizardTargetWeightResult {
    #[serde(deserialize_with = "crate::crud::serde::date::deserialize")]
    date_per_rate: HashMap<i32, String>,
    target_classification: BmiCategory,
    warning: bool,
    message: String,
}

#[derive(Deserialize, Debug)]
#[serde(rename_all = "camelCase")]
pub struct WizardTargetDateInput {
    age: i32,
    sex: CalculationSex,
    current_weight: f32,
    height: f32,
    calculation_goal: CalculationGoal,
    #[serde(deserialize_with = "crate::crud::serde::date::deserialize")]
    target_date: NaiveDate,
}

#[derive(Serialize, Debug)]
pub struct WizardTargetDateResult {
    result_by_rate: HashMap<i32, WizardResult>,
}

#[derive(Serialize, Deserialize, Debug)]
enum CalculationGoal {
    GAIN,
    LOSS,
}

#[derive(Serialize, Deserialize, Debug)]
enum CalculationSex {
    MALE,
    FEMALE,
}

#[derive(Serialize, Deserialize, Clone, Debug)]
#[derive(PartialEq)]
enum BmiCategory {
    UNDERWEIGHT,
    STANDARD_WEIGHT,
    OVERWEIGHT,
    OBESE,
    SEVERELY_OBESE,
}

#[derive(Serialize, Deserialize, Debug)]
enum WizardRecommendation {
    HOLD,
    LOSE,
    GAIN,
}

#[derive(Serialize, Deserialize)]
struct ErrorDescription {
    property_name: String,
    message: String,
}

fn validate_activity_level(activity_level: f32) -> Result<(), ValidationError> {
    match activity_level {
        1f32 | 1.25f32 | 1.5f32 | 1.75f32 | 2f32 => Ok(()),
        _ => Err(ValidationError::new("Please enter a valid activity level.")),
    }
}

pub fn calculate(wizard_input: WizardInput) -> WizardResult {
    let target_bmi_range = calculate_target_bmi(&wizard_input.age);

    let bmr = calculate_bmr(
        &wizard_input.sex,
        &wizard_input.weight,
        &wizard_input.height,
        &wizard_input.age,
    );

    let tdee = calculate_tdee(&wizard_input.activity_level, &bmr);
    let deficit = calculate_deficit(&wizard_input.weekly_difference);
    let target = calculate_target(&wizard_input.calculation_goal, &tdee, &deficit);

    let bmi = calculate_bmi(&wizard_input.weight, &wizard_input.height);
    let bmi_category = calculate_bmi_category(&wizard_input.sex, &bmi);

    let target_bmi_lower = target_bmi_range.start().clone();
    let target_bmi_upper = target_bmi_range.end().clone();
    let target_bmi = (target_bmi_upper + target_bmi_lower) / 2;

    let target_weight = calculate_target_weight(&target_bmi_range, &wizard_input.height);
    let target_weight_lower =
        calculate_target_weight_lower(&target_bmi_range, &wizard_input.height);
    let target_weight_upper =
        calculate_target_weight_upper(&target_bmi_range, &wizard_input.height);

    let recommendation = calculate_recommendation(&bmi_category);

    let mut wizard_result = WizardResult {
        bmr,
        tdee,
        deficit,
        target,
        bmi,
        bmi_category,
        recommendation,
        target_bmi,
        target_bmi_upper,
        target_bmi_lower,
        target_weight,
        target_weight_upper,
        target_weight_lower,
        duration_days: 0,
        duration_days_upper: 0,
        duration_days_lower: 0,
    };

    calculate_duration_days_total(&wizard_input, &mut wizard_result);

    wizard_result
}

pub fn calculate_for_target_date(input: &WizardTargetDateInput) -> WizardTargetDateResult {
    let today = Utc::now().date_naive();

    let days_between = (input.target_date - today).num_days() as i32;

    let current_bmi = calculate_bmi(&input.current_weight, &input.height);
    let current_classification = calculate_bmi_category(&input.sex, &current_bmi);

    let rates: Vec<i32> = vec![100, 200, 300, 400, 500, 600, 700];
    let multiplier = if matches!(input.calculation_goal, CalculationGoal::LOSS) { -1 } else { 1 };
    let obese_list = vec![BmiCategory::OBESE, BmiCategory::SEVERELY_OBESE];

    let mut result_by_rate: HashMap<i32, WizardResult> = rates.into_iter().filter_map(|rate| {
        let weight = input.current_weight + ((multiplier * days_between * rate) as f32) / 7000.0;

        let result_bmi = calculate_bmi(&weight, &input.height);
        let result_bmi_category = calculate_bmi_category(&input.sex, &result_bmi);

        let result = WizardResult::specific(&result_bmi, result_bmi_category.clone(), &weight);

        match input.calculation_goal {
            CalculationGoal::LOSS => {
                if result_bmi_category != BmiCategory::UNDERWEIGHT {
                    Some((rate, result))
                } else {
                    None
                }
            }
            _ => {
                if !obese_list.contains(&result_bmi_category) || obese_list.contains(&current_classification) {
                    Some((rate, result))
                } else {
                    None
                }
            }
        }
    }).collect();

    // calculate a custom rate to present at least one viable result
    if result_by_rate.is_empty() {
        let target_bmi = calculate_target_bmi(&input.age);
        let target_weight = calculate_target_weight(&target_bmi, &input.height);

        let difference = match input.calculation_goal {
            CalculationGoal::GAIN => {
                target_weight - input.current_weight
            }
            CalculationGoal::LOSS => {
                input.current_weight - target_weight
            }
        }.round() as i32;

        let rate = (difference as f32 * 7000.0 / (days_between as f32)).round() as i32;

        let result_bmi = calculate_bmi(&target_weight, &input.height);
        let result_bmi_category = calculate_bmi_category(&input.sex, &result_bmi);

        let result = WizardResult::specific(&result_bmi, result_bmi_category.clone(), &target_weight);
        
        result_by_rate.insert(rate, result);
    }
    
    WizardTargetDateResult {
        result_by_rate
    }
}

pub fn calculate_for_target_weight(input: &WizardTargetWeightInput) -> WizardTargetWeightResult {
    let current_bmi = calculate_bmi(&input.current_weight, &input.height);
    let target_bmi = calculate_bmi(&input.target_weight, &input.height);

    let current_classification = calculate_bmi_category(&input.sex, &current_bmi);
    let target_classification = calculate_bmi_category(&input.sex, &target_bmi);

    let difference = if input.target_weight > input.current_weight {
        input.target_weight - input.current_weight
    } else {
        input.current_weight - input.target_weight
    };

    let warning = matches!(target_classification, BmiCategory::UNDERWEIGHT | BmiCategory::OBESE | BmiCategory::SEVERELY_OBESE);
    let mut message = String::new();

    if let BmiCategory::UNDERWEIGHT = target_classification {
        if input.target_weight < input.current_weight {
            if current_classification == BmiCategory::UNDERWEIGHT {
                message = "Your target weight is even lower than your current weight...".to_string();
            } else {
                message = "Your target weight will classify you as underweight...".to_string();
            }
        }
    } else if let BmiCategory::OBESE = target_classification {
        if input.target_weight > input.current_weight {
            if current_classification == BmiCategory::OBESE {
                message = "Your target weight is even higher than your current weight...".to_string();
            } else {
                message = "Your target weight will classify you as obese...".to_string();
            }
        }
    } else if let BmiCategory::SEVERELY_OBESE = target_classification {
        if input.target_weight > input.current_weight {
            if current_classification == BmiCategory::SEVERELY_OBESE {
                message = "Your target weight is even higher than your current weight...".to_string();
            } else {
                message = "Your target weight will classify you as severely obese...".to_string();
            }
        }
    }

    let mut date_per_rate = HashMap::new();

    if message.is_empty() {
        let rates = vec![100, 200, 300, 400, 500, 600, 700];
        for rate in rates {
            let days = (difference * 7000.0 / rate as f32).round() as i64;
            date_per_rate.insert(rate, (input.start_date + Duration::days(days)).to_string());
        }
    }

    let result = WizardTargetWeightResult {
        date_per_rate,
        target_classification,
        warning,
        message,
    };

    result
}

fn calculate_bmr(sex: &CalculationSex, weight: &f32, height: &f32, age: &i32) -> f32 {
    match sex {
        CalculationSex::MALE => (66.0 + 13.7 * weight + 5.0 * height - 6.8 * (*age as f32)).round(),
        CalculationSex::FEMALE => {
            (655.0 + 9.6 * weight + 1.8 * height - 4.7 * (*age as f32)).round()
        }
    }
}

fn calculate_tdee(activity_level: &f32, bmr: &f32) -> f32 {
    (activity_level * bmr).round()
}

fn calculate_deficit(weekly_difference: &i32) -> f32 {
    (weekly_difference.clone() as f32 / 10.0 * 7000.0 / 7.0).round()
}

fn calculate_target(calculation_goal: &CalculationGoal, tdee: &f32, deficit: &f32) -> f32 {
    match calculation_goal {
        CalculationGoal::GAIN => tdee + deficit,
        CalculationGoal::LOSS => tdee - deficit,
    }
}

fn calculate_bmi(weight: &f32, height: &f32) -> f32 {
    (weight / ((height / 100.0).powi(2))).round()
}

fn calculate_bmi_category(sex: &CalculationSex, bmi: &f32) -> BmiCategory {
    match sex {
        CalculationSex::FEMALE => match bmi {
            0.0..18.0 => BmiCategory::UNDERWEIGHT,
            19.0..24.0 => BmiCategory::STANDARD_WEIGHT,
            25.0..30.0 => BmiCategory::OVERWEIGHT,
            31.0..40.0 => BmiCategory::OBESE,
            _ => BmiCategory::SEVERELY_OBESE,
        },
        CalculationSex::MALE => match bmi {
            0.0..19.0 => BmiCategory::UNDERWEIGHT,
            20.0..25.0 => BmiCategory::STANDARD_WEIGHT,
            26.0..30.0 => BmiCategory::OVERWEIGHT,
            31.0..40.0 => BmiCategory::OBESE,
            _ => BmiCategory::SEVERELY_OBESE,
        },
    }
}

fn calculate_target_bmi(age: &i32) -> std::ops::RangeInclusive<i32> {
    match age {
        19..=24 => 19..=24,
        25..=34 => 20..=25,
        35..=44 => 21..=26,
        45..=54 => 22..=27,
        55..=64 => 23..=28,
        _ => 24..=29,
    }
}

fn calculate_target_weight(target_bmi: &std::ops::RangeInclusive<i32>, height: &f32) -> f32 {
    (((target_bmi.start() + target_bmi.end()) as f32 / 2.0) * (height / 100.0).powi(2)).round()
}

fn calculate_target_weight_lower(target_bmi: &std::ops::RangeInclusive<i32>, height: &f32) -> f32 {
    (*(target_bmi.start()) as f32 * (height / 100.0).powi(2)).round()
}

fn calculate_target_weight_upper(target_bmi: &std::ops::RangeInclusive<i32>, height: &f32) -> f32 {
    (*(target_bmi.end()) as f32 * (height / 100.0).powi(2)).round()
}

fn calculate_duration_days_total(wizard_input: &WizardInput, wizard_result: &mut WizardResult) {
    match wizard_input.calculation_goal {
        CalculationGoal::GAIN => {
            wizard_result.duration_days = calculate_duration_days(
                wizard_result.target_weight,
                wizard_input.weight,
                wizard_result.deficit,
            );
            wizard_result.duration_days_upper = calculate_duration_days(
                wizard_result.target_weight_upper,
                wizard_input.weight,
                wizard_result.deficit,
            );
            wizard_result.duration_days_lower = calculate_duration_days(
                wizard_result.target_weight_lower,
                wizard_input.weight,
                wizard_result.deficit,
            );
        }
        CalculationGoal::LOSS => {
            wizard_result.duration_days = calculate_duration_days(
                wizard_input.weight,
                wizard_result.target_weight,
                wizard_result.deficit,
            );
            wizard_result.duration_days_upper = calculate_duration_days(
                wizard_input.weight,
                wizard_result.target_weight_upper,
                wizard_result.deficit,
            );
            wizard_result.duration_days_lower = calculate_duration_days(
                wizard_input.weight,
                wizard_result.target_weight_lower,
                wizard_result.deficit,
            );
        }
    }
}

fn calculate_duration_days(weight: f32, target_weight: f32, deficit: f32) -> i32 {
    ((weight - target_weight) * 7000.0 / deficit).floor() as i32
}

fn calculate_recommendation(bmi_category: &BmiCategory) -> WizardRecommendation {
    match bmi_category {
        BmiCategory::STANDARD_WEIGHT => WizardRecommendation::HOLD,
        BmiCategory::UNDERWEIGHT => WizardRecommendation::GAIN,
        BmiCategory::OVERWEIGHT => WizardRecommendation::LOSE,
        BmiCategory::OBESE => WizardRecommendation::LOSE,
        BmiCategory::SEVERELY_OBESE => WizardRecommendation::LOSE,
    }
}
